<!DOCTYPE html>
<html lang="es">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>

    <script>
       function sumar(a,b,callback)
        {
            return new Promise((resolve,reject)=>{
                let suma;
                setTimeout(() => {
                    if(typeof a !== "number" || typeof b !== "number")
                    {
                        //esto hago sino puede obtener y quiero que devuelva que esta todo mal
                        reject({mensaje: "Paramentros no numericos en sumar"});
                        //entre llaves es un objeto literal
                    }
                    else
                    {
                        suma = a+b;
                        console.log(suma);
                        console.log("Tengo el resultado de la suma!!!");
                        resolve(suma);//si esta todo bien se lo paso a resolve
                    }
                    
                }, 2000);        
            });
                
        }
        function cuadrado(a)
        {
            return new Promise((resolve,reject)=>{
                let cuad;
                setTimeout(() => {
                    if(typeof a !== "number")
                    {
                        reject({mensaje: "Paramentros no numericos en cuadrado"});
                    }
                    else
                    {
                        cuad = a*a;
                        console.log(cuad);
                        console.log("Tengo el resultado del cuadrado!!!");
                        resolve(cuad);
                    }
                    
                }, 1500);
            });
           
        }
        function multiplicar(a,b)
        {
            //retorna esta promesa
            return new Promise((resolve,reject)=>{
                let producto;
                setTimeout(() => {
                    if(typeof a !== "number" || typeof b !== "number")
                    {
                        //esto hago sino puede obtener y quiero que devuelva que esta todo mal
                        reject({mensaje: "Paramentros no numericos en multiplicar"});
                    }
                    else
                    {
                        producto = a+b;
                        console.log(producto);
                        console.log("Tengo el resultado del producto!!!");
                        resolve(producto);//si esta todo bien se lo paso a resolve
                    }
                }, 2000);        
            });
        }

        //esto no funciona porque es un funcion sincronica
        /*function realizarOperaciones()
        {
            let suma = sumar(4,3);
            let cuad = cuadrado(suma);
            let rta = multiplicar(cuad,10);
            console.log(rta);
        }*/
        
        //para que sea asincrona le tengo que poner la palabra async en la declaracion y await adelante de la llamada de las funciones
        async function realizarOperaciones()
        {
            //esto que era asincronico lo vuelvo a hacer sincronico
            //y lo coloco en un try catch para detectar si alguna promesa no se cumplio
            try {
                let suma = await sumar(4,3);
                let cuad = await cuadrado(suma);
                let rta = await multiplicar(cuad,10);
                console.log(rta);
            } catch (error) {
                console.log(error.mensaje);
            }
            
        }
        realizarOperaciones();
        //Con promesa:
        /*
        sumar(4,3)
        .then((suma)=>cuadrado(suma))
        .then((cuad)=>multiplicar(cuad,10))
        .then((producto)=>{
            console.log(producto);
        })//sino le pongo punto y coma puedo agarrar en un catch lo que se hizo en el reject
        .catch((error)=>{
            console.log(error.mensaje)
        });*/
        

    </script>
    
</body>
</html>